<a href="https://colab.research.google.com/github/dacardonave/Modelo_Desercion_Clientes_De_Una_Empresa/blob/main/Modelo_Desercion_Clientes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Prueba tecnica Ingeniero de Datos – Comercial Nutresa 2021

### Daniel Cardona Velasquez

## Planteamiento del reto

###### La compañía para la cual trabaja desea contar con una herramienta que le permita identificar anticipadamente los clientes que se van a retirar en el siguiente periodo, con el objetivo de implementar planes de acción que logren reducir este comportamiento tanto como sea posible. Por este motivo, se le ha encargado construir un modelo que a partir de la base de datos suministrada permita predecir, con la mayor precisión posible, la variable “Retiro”.

###### Para cumplir con este reto debe dar respuesta a las preguntas y lineamientos que se indican a continuación:

* Realice un análisis descriptivo de los datos, depure y seleccione las variables que considere más importantes.
*	Defina que indicador debe usar para medir la precisión del modelo y la metodología de validación a seguir.
*	Entrene y valide los modelos que considere necesarios y seleccione el mejor con base en sus criterios.
*	Plantee un plan de trabajo de puesta en marcha y administración del modelo.
*	Envíe un archivo Markdown o Jupyter Notebook que responda a cada uno de los puntos anteriores, debidamente argumentado y en el cual se explique el paso a paso realizado.


### Librerias

In [2]:
 #Para admitir python 2 y python 3
from __future__ import division, print_function, unicode_literals

# Importo las librerias a usar
import pandas as pd
import sklearn as sk
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
import pickle as pk
import numpy as np
from sklearn.metrics import confusion_matrix
import joblib
import os
from joblib import dump
from sklearn.naive_bayes import GaussianNB

#Ignora las advertencias inútiles
import warnings
warnings.filterwarnings(action="ignore", message="^internal gelsd")

# plot
%matplotlib inline
import matplotlib as mpl
import matplotlib.pyplot as plt
from matplotlib import colors

### Importacion de los datos y lectura de los datos


In [17]:
BD = pd.read_csv('Data.csv', sep=";")

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (14,16,20,21,24,25,28,29,36,42,45,47,55,56,57,62) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


## Datos 

Se observan 63 variables y la etiqueta de salida (Retiro) a predecir

In [23]:
BD.head(10)

Id  Lifetime  ...  last_100_gprs_usage  Retiro
0  13      1000  ...                    0       0
1  14      1000  ...                    0       0
2  18      1005  ...                    0       0
3  27      1013  ...                    0       1
4  32      1032  ...                    0       0
5  35      1039  ...                    0       0
6  36      1040  ...                    0       0
7  47       521  ...                    0       0
8  52       824  ...                 2,77       0
9  59       867  ...                    0       0

[10 rows x 64 columns]

## Análisis descriptivo y exploratorio


In [35]:
# Tamaño del dataset
print(BD.shape)

(66469, 64)


In [28]:
# Informacion sobre tipo y cantidad de los datos
BD.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66469 entries, 0 to 66468
Data columns (total 64 columns):
 #   Column                                      Non-Null Count  Dtype 
---  ------                                      --------------  ----- 
 0   Id                                          66469 non-null  int64 
 1   Lifetime                                    66469 non-null  int64 
 2   user_intake                                 66469 non-null  int64 
 3   user_no_outgoing_activity_in_days           66469 non-null  int64 
 4   user_account_balance_last                   66469 non-null  object
 5   user_spendings                              66469 non-null  object
 6   user_has_outgoing_calls                     66469 non-null  int64 
 7   user_has_outgoing_sms                       66469 non-null  int64 
 8   user_use_gprs                               66469 non-null  int64 
 9   user_does_reload                            66469 non-null  int64 
 10  reloads_inactive_days 

In [33]:
# Estadisticas basica de las vairbales
BD.describe()

Id  ...        Retiro
count  6.646900e+04  ...  66469.000000
mean   5.948572e+05  ...      0.209225
std    4.292423e+05  ...      0.406759
min    1.300000e+01  ...      0.000000
25%    2.501070e+05  ...      0.000000
50%    4.135120e+05  ...      0.000000
75%    1.041120e+06  ...      0.000000
max    1.196944e+06  ...      1.000000

[8 rows x 36 columns]

In [38]:
#conteo de los valores unicos de la etiqueta a predecir 
BD["Retiro"].value_counts() 

0    52562
1    13907
Name: Retiro, dtype: int64

# Matriz de Correlacion

La correlacion permite hacer inferencias a partir de los datos.
Indica la relacion lineal entre los parametros. Un valor de correlación alto y positivo indica que los elementos miden la misma característica. Si los elementos no están altamente correlacionados, entonces los elementos pueden medir diferentes características.

In [39]:
corr_matrix = BD.corr()
corr_matrix.style.background_gradient(cmap='coolwarm')

La columna de salida "Retiro" es la que se quiere predecir, por ende ordenamos los valores de Correlacion entre la salida y las demas variables

In [40]:
corr_matrix["Retiro"].sort_values(ascending=False)

Retiro                                    1.000000
calls_outgoing_to_offnet_inactive_days    0.546173
calls_outgoing_inactive_days              0.546173
calls_outgoing_to_onnet_inactive_days     0.546173
calls_outgoing_to_abroad_inactive_days    0.546173
sms_outgoing_to_abroad_inactive_days      0.540235
sms_outgoing_to_offnet_inactive_days      0.540235
sms_outgoing_to_onnet_inactive_days       0.540235
sms_outgoing_inactive_days                0.540235
Lifetime                                  0.529719
Id                                        0.357816
user_no_outgoing_activity_in_days         0.222895
reloads_inactive_days                     0.132148
gprs_inactive_days                        0.128939
user_intake                              -0.014725
gprs_session_count                       -0.022318
calls_outgoing_to_abroad_count           -0.040320
user_does_reload                         -0.043394
last_100_sms_outgoing_to_abroad_count    -0.051195
calls_outgoing_to_onnet_count  

En la celda anterior podemos ver la correlacion de todas la variables del dataser con la variable etiqueta (Retiro) que se busca predecir con el modelo. Una correlacion positiva nos indica que esas dos variables aumentan o disminuyen en una misma direccion, mientras que una correlacion negativa indica que mientras una aumenta la otra disminuye. 

Con base en esto y por pura inferencia estadistica podriamos decir que las variables que tienen mas influencia sobre la desercion de los clientes son las siguientes:

* user_has_outgoing_calls       --->           -0.553790
* user_has_outgoing_sms                    -0.490214
* calls_outgoing_to_offnet_inactive_days --->   0.546173
* calls_outgoing_inactive_days       --->       0.546173
* calls_outgoing_to_onnet_inactive_days  --->   0.546173
* calls_outgoing_to_abroad_inactive_days  --->  0.546173
* sms_outgoing_to_abroad_inactive_days  --->    0.540235
* sms_outgoing_to_offnet_inactive_days   --->   0.540235
* sms_outgoing_to_onnet_inactive_days   --->    0.540235
* sms_outgoing_inactive_days        --->        0.540235
* Lifetime                         --->         0.529719
